File: src/expressions.py
"""
expressions.py — Infix <-> Postfix conversions (with file helpers)

- infix_to_postfix(expr: str) -> str
- postfix_to_infix(expr: str) -> str
Both handle unary minus by representing it as "~" in postfix form.

File helpers:
- convert_file(mode, in_path, out_path)
  mode in {"infix2postfix", "postfix2infix"}
  Reads one expression per line and writes converted form per line.
"""
from typing import List

# Precedence and associativity
precedence = {'~': 5, '^': 4, '*': 3, '/': 3, '+': 2, '-': 2}
right_associative = {'^', '~'}

def is_operator(tok: str) -> bool:
    return tok in precedence

def _tokenize(expr: str) -> List[str]:
    import re
    token_specification = [
        ('NUMBER',   r'\d+(\.\d+)?'),
        ('ID',       r'[A-Za-z_]\w*'),
        ('OP',       r'[\+\-\*\/\^\(\)]'),
        ('SKIP',     r'[ \t]+'),
        ('MISMATCH', r'.'),
    ]
    tok_regex = '|'.join('(?P<%s>%s)' % pair for pair in token_specification)
    tokens = []
    for mo in re.finditer(tok_regex, expr):
        kind = mo.lastgroup
        value = mo.group()
        if kind in ('NUMBER', 'ID', 'OP'):
            tokens.append(value)
        elif kind == 'SKIP':
            continue
        else:
            raise ValueError(f'Unexpected character {value!r} in expression: {expr}')
    return tokens

def infix_to_postfix(expr: str) -> str:
    tokens = _tokenize(expr)
    out: List[str] = []
    stack: List[str] = []  # operator stack
    prev_tok_type = None
    for tok in tokens:
        if tok.isdigit() or tok.replace('.', '', 1).isdigit() or tok[0].isalpha() or tok[0] == '_':
            out.append(tok)
            prev_tok_type = 'OPERAND'
        elif tok in ('+', '-', '*', '/', '^'):
            # detect unary minus
            o1 = '~' if tok == '-' and (prev_tok_type is None or prev_tok_type in ('OP', 'LPAREN')) else tok
            while stack and stack[-1] != '(' and (precedence[stack[-1]] > precedence[o1] or
                                                  (precedence[stack[-1]] == precedence[o1] and o1 not in right_associative)):
                out.append(stack.pop())
            stack.append(o1)
            prev_tok_type = 'OP'
        elif tok == '(':
            stack.append('(')
            prev_tok_type = 'LPAREN'
        elif tok == ')':
            while stack and stack[-1] != '(':
                out.append(stack.pop())
            if not stack:
                raise ValueError("Mismatched parentheses")
            stack.pop()  # '('
            prev_tok_type = 'RPAREN'
        else:
            raise ValueError(f"Unknown token: {tok}")
    while stack:
        if stack[-1] in ('(', ')'):
            raise ValueError("Mismatched parentheses")
        out.append(stack.pop())
    return ' '.join(out)

class _Node:
    __slots__ = ('expr','prec','arity')
    def __init__(self, expr: str, prec: int, arity: int = 0):
        self.expr = expr
        self.prec = prec
        self.arity = arity

def _maybe_paren(child: '_Node', parent_prec: int, is_right_child: bool, op: str) -> str:
    if child.arity == 0:
        return child.expr
    need = False
    if child.prec < parent_prec:
        need = True
    elif child.prec == parent_prec:
        if op in ('-','/'):
            if is_right_child:
                need = True
        elif op == '^':
            if not is_right_child:
                need = True
    return f'({child.expr})' if need else child.expr

def postfix_to_infix(postfix: str) -> str:
    tokens = postfix.split()
    stack: List[_Node] = []
    for tok in tokens:
        if is_operator(tok):
            if tok == '~':
                if len(stack) < 1:
                    raise ValueError("Insufficient operands for unary operator '~'")
                a = stack.pop()
                expr = f'-{_maybe_paren(a, precedence[tok], True, tok)}'
                stack.append(_Node(expr, precedence[tok], arity=1))
            else:
                if len(stack) < 2:
                    raise ValueError(f"Insufficient operands for operator {tok}")
                b = stack.pop()
                a = stack.pop()
                a_expr = _maybe_paren(a, precedence[tok], False, tok)
                b_expr = _maybe_paren(b, precedence[tok], True, tok)
                expr = f'{a_expr} {tok} {b_expr}'
                stack.append(_Node(expr, precedence[tok], arity=2))
        else:
            stack.append(_Node(tok, 100, arity=0))
    if len(stack) != 1:
        raise ValueError("Invalid postfix expression")
    return stack[0].expr

# ---------- File helpers ----------
def convert_file(mode: str, in_path: str, out_path: str):
    """Convert lines in a file.
    mode: 'infix2postfix' or 'postfix2infix'
    Each line is treated as a separate expression.
    """
    mode = mode.lower()
    if mode not in ('infix2postfix', 'postfix2infix'):
        raise ValueError("mode must be 'infix2postfix' or 'postfix2infix'")
    out_lines = []
    with open(in_path, 'r', encoding='utf-8') as fin:
        for line_no, line in enumerate(fin, 1):
            s = line.strip()
            if not s:
                out_lines.append('')
                continue
            try:
                if mode == 'infix2postfix':
                    out_lines.append(infix_to_postfix(s))
                else:
                    out_lines.append(postfix_to_infix(s))
            except Exception as e:
                out_lines.append(f'# Error on line {line_no}: {e}')
    with open(out_path, 'w', encoding='utf-8') as fout:
        fout.write('\n'.join(out_lines))

if __name__ == "__main__":
    # demo
    print("Demo infix -> postfix")
    for t in ["A * ( B + C ) - D / E", "3 + 4 * 2 / ( 1 - 5 ) ^ 2 ^ 3", "-A + B"]:
        print(t, "->", infix_to_postfix(t))
    print("\nDemo postfix -> infix")
    for t in ["A B C + * D E / -", "3 4 2 * 1 5 - 2 3 ^ ^ / +", "A ~ B +"]:
        print(t, "->", postfix_to_infix(t))

File: src/Expressions.java
/*
expressions.py — Infix <-> Postfix conversions (with file helpers)

- infix_to_postfix(expr: str) -> str
- postfix_to_infix(expr: str) -> str
Both handle unary minus by representing it as "~" in postfix form.

File helpers:
- convert_file(mode, in_path, out_path)
  mode in {"infix2postfix", "postfix2infix"}
  Reads one expression per line and writes converted form per line.
*/

// Expressions.java - Convert between infix and postfix
import java.util.ArrayList;
import java.util.Stack;
import java.io.*;

public class Expressions {

    // Check if character is operator
    private static boolean isOperator(String op) {
        return op.equals("+") || op.equals("-") || op.equals("*") ||
                op.equals("/") || op.equals("^") || op.equals("~");
    }

    // Get operator priority
    private static int getPriority(String op) {
        if (op.equals("~"))
            return 5; // unary minus
        if (op.equals("^"))
            return 4;
        if (op.equals("*") || op.equals("/"))
            return 3;
        if (op.equals("+") || op.equals("-"))
            return 2;
        return 0;
    }

    // Check if operator is right associative
    private static boolean isRightAssociative(String op) {
        return op.equals("^") || op.equals("~");
    }

    // Split expression into tokens (numbers, letters, operators)
    private static ArrayList<String> tokenize(String expr) {
        ArrayList<String> tokens = new ArrayList<>();
        String current = "";

        for (int i = 0; i < expr.length(); i++) {
            char ch = expr.charAt(i);

            if (ch == ' ' || ch == '\t') {
                // skip spaces
                continue;
            } else if (Character.isLetterOrDigit(ch) || ch == '_' || ch == '.') {
                // build number or variable name
                current += ch;
            } else {
                // operator or parenthesis
                if (!current.isEmpty()) {
                    tokens.add(current);
                    current = "";
                }
                tokens.add(String.valueOf(ch));
            }
        }
        if (!current.isEmpty()) {
            tokens.add(current);
        }
        return tokens;
    }

    // Convert INFIX to POSTFIX
    public static String infixToPostfix(String infix) {
        ArrayList<String> tokens = tokenize(infix);
        ArrayList<String> output = new ArrayList<>();
        Stack<String> stack = new Stack<>();
        String lastType = null;

        for (String token : tokens) {
            // Check if it's a number or variable
            if (Character.isLetterOrDigit(token.charAt(0))) {
                output.add(token);
                lastType = "OPERAND";
            }
            // Check for operators
            else if (token.equals("+") || token.equals("-") ||
                    token.equals("*") || token.equals("/") || token.equals("^")) {

                // Check if minus is unary (like -5 or -A)
                String op = token;
                if (token.equals("-") && (lastType == null ||
                        lastType.equals("OP") || lastType.equals("("))) {
                    op = "~"; // special symbol for unary minus
                }

                // Pop operators with higher priority
                while (!stack.isEmpty() && !stack.peek().equals("(")) {
                    String top = stack.peek();
                    int topPriority = getPriority(top);
                    int opPriority = getPriority(op);

                    if (topPriority > opPriority ||
                            (topPriority == opPriority && !isRightAssociative(op))) {
                        output.add(stack.pop());
                    } else {
                        break;
                    }
                }
                stack.push(op);
                lastType = "OP";
            }
            // Left parenthesis
            else if (token.equals("(")) {
                stack.push("(");
                lastType = "(";
            }
            // Right parenthesis
            else if (token.equals(")")) {
                while (!stack.isEmpty() && !stack.peek().equals("(")) {
                    output.add(stack.pop());
                }
                if (stack.isEmpty()) {
                    throw new RuntimeException("Mismatched parentheses");
                }
                stack.pop(); // remove '('
                lastType = ")";
            }
        }

        // Pop remaining operators
        while (!stack.isEmpty()) {
            if (stack.peek().equals("(") || stack.peek().equals(")")) {
                throw new RuntimeException("Mismatched parentheses");
            }
            output.add(stack.pop());
        }

        // Join with spaces
        String result = "";
        for (int i = 0; i < output.size(); i++) {
            result += output.get(i);
            if (i < output.size() - 1) {
                result += " ";
            }
        }
        return result;
    }

    // Helper class for building infix expression
    private static class ExprNode {
        String expr;
        int priority;
        int operandCount; // 0 for operand, 1 for unary, 2 for binary

        ExprNode(String expr, int priority, int operandCount) {
            this.expr = expr;
            this.priority = priority;
            this.operandCount = operandCount;
        }
    }

    // Add parentheses if needed
    private static String addParens(ExprNode child, int parentPriority,
            boolean isRightSide, String op) {
        if (child.operandCount == 0) {
            return child.expr; // operand, no parens needed
        }

        boolean needParens = false;
        if (child.priority < parentPriority) {
            needParens = true;
        } else if (child.priority == parentPriority) {
            if ((op.equals("-") || op.equals("/")) && isRightSide) {
                needParens = true;
            } else if (op.equals("^") && !isRightSide) {
                needParens = true;
            }
        }

        if (needParens) {
            return "(" + child.expr + ")";
        }
        return child.expr;
    }

    // Convert POSTFIX to INFIX
    public static String postfixToInfix(String postfix) {
        String[] tokens = postfix.split("\\s+");
        Stack<ExprNode> stack = new Stack<>();

        for (String token : tokens) {
            if (isOperator(token)) {
                if (token.equals("~")) {
                    // Unary minus
                    if (stack.isEmpty()) {
                        throw new RuntimeException("Not enough operands for ~");
                    }
                    ExprNode operand = stack.pop();
                    String expr = "-" + addParens(operand, getPriority(token), true, token);
                    stack.push(new ExprNode(expr, getPriority(token), 1));
                } else {
                    // Binary operator
                    if (stack.size() < 2) {
                        throw new RuntimeException("Not enough operands for " + token);
                    }
                    ExprNode right = stack.pop();
                    ExprNode left = stack.pop();
                    String leftExpr = addParens(left, getPriority(token), false, token);
                    String rightExpr = addParens(right, getPriority(token), true, token);
                    String expr = leftExpr + " " + token + " " + rightExpr;
                    stack.push(new ExprNode(expr, getPriority(token), 2));
                }
            } else {
                // Operand (number or variable)
                stack.push(new ExprNode(token, 100, 0));
            }
        }

        if (stack.size() != 1) {
            throw new RuntimeException("Invalid postfix expression");
        }
        return stack.pop().expr;
    }

    // Convert file (one expression per line)
    public static void convertFile(String mode, String inputFile, String outputFile) {
        try {
            BufferedReader reader = new BufferedReader(new FileReader(inputFile));
            BufferedWriter writer = new BufferedWriter(new FileWriter(outputFile));
            String line;
            int lineNum = 1;

            while ((line = reader.readLine()) != null) {
                line = line.trim();
                if (line.isEmpty()) {
                    writer.write("");
                    writer.newLine();
                    continue;
                }

                try {
                    String result;
                    if (mode.equals("infix2postfix")) {
                        result = infixToPostfix(line);
                    } else {
                        result = postfixToInfix(line);
                    }
                    writer.write(result);
                } catch (Exception e) {
                    writer.write("# Error on line " + lineNum + ": " + e.getMessage());
                }
                writer.newLine();
                lineNum++;
            }

            reader.close();
            writer.close();
        } catch (IOException e) {
            System.out.println("Error: " + e.getMessage());
        }
    }

    // Main - demo
    public static void main(String[] args) {
        System.out.println("Demo infix -> postfix");
        System.out.println("A * ( B + C ) - D / E -> " +
                infixToPostfix("A * ( B + C ) - D / E"));
        System.out.println("3 + 4 * 2 / ( 1 - 5 ) ^ 2 ^ 3 -> " +
                infixToPostfix("3 + 4 * 2 / ( 1 - 5 ) ^ 2 ^ 3"));
        System.out.println("-A + B -> " + infixToPostfix("-A + B"));

        System.out.println("\nDemo postfix -> infix");
        System.out.println("A B C + * D E / - -> " +
                postfixToInfix("A B C + * D E / -"));
        System.out.println("3 4 2 * 1 5 - 2 3 ^ ^ / + -> " +
                postfixToInfix("3 4 2 * 1 5 - 2 3 ^ ^ / +"));
        System.out.println("A ~ B + -> " + postfixToInfix("A ~ B +"));
    }
}